// ai_service.proto
// [Core] gRPC 定义：StreamPredict、ControlMessage、Backpressure、Training
// 说明：此文件用于定义双向流预测接口、控制消息、背压信号和训练接口的 Protobuf schema。

syntax = "proto3";

package cy.llm;

option java_multiple_files = true;
option java_package = "com.cy.llm.proto";
option java_outer_classname = "AiServiceProto";

// ========== 公共消息定义 ==========

message StreamMetadata {
	string trace_id = 1;
	string tenant = 2;
	string player_id = 3;
	string locale = 4;
	map<string, string> extra = 5;
}

message GenerationParameters {
	uint32 max_new_tokens = 1;
	double temperature = 2;
	double top_p = 3;
	double repetition_penalty = 4;
}

// ========== 推理服务消息 ==========

message StreamPredictRequest {
	string model_id = 1;
	string prompt = 2;
	string adapter = 3;
	int32 priority = 4;
	GenerationParameters generation = 5;
	StreamMetadata metadata = 6;
	string worker_hint = 7;
}

message StreamPredictResponse {
	string trace_id = 1;
	string chunk = 2;
	bool end_of_stream = 3;
	uint32 index = 4;
}

message ControlMessage {
	string trace_id = 1;
	string command = 2;
	map<string, string> payload = 3;
}

message WorkerHealthRequest {
	string trace_id = 1;
}

message WorkerHealthResponse {
	bool healthy = 1;
	map<string, string> metrics = 2;
}

// ========== 训练服务消息 ==========

// 训练任务状态
enum TrainingStatus {
	TRAINING_STATUS_UNKNOWN = 0;
	TRAINING_STATUS_QUEUED = 1;
	TRAINING_STATUS_RUNNING = 2;
	TRAINING_STATUS_COMPLETED = 3;
	TRAINING_STATUS_FAILED = 4;
	TRAINING_STATUS_CANCELLED = 5;
}

// LoRA 训练配置
message LoraConfig {
	uint32 r = 1;              // LoRA rank，默认 64
	uint32 lora_alpha = 2;     // LoRA alpha，默认 16
	double lora_dropout = 3;   // Dropout 率，默认 0.05
	repeated string target_modules = 4;  // 目标模块，默认 ["q_proj", "v_proj"]
}

// 量化配置
message QuantizationConfig {
	bool use_4bit = 1;         // 是否使用 4-bit 量化
	string bnb_4bit_compute_dtype = 2;  // 计算精度，默认 "float16"
	string bnb_4bit_quant_type = 3;     // 量化类型，默认 "nf4"
}

// 训练超参数
message TrainingHyperparams {
	uint32 num_train_epochs = 1;    // 训练轮数，默认 3
	uint32 per_device_batch_size = 2;  // 每设备批量大小，默认 4
	uint32 gradient_accumulation_steps = 3;  // 梯度累积步数，默认 4
	double learning_rate = 4;       // 学习率，默认 2e-4
	double warmup_ratio = 5;        // 预热比例，默认 0.03
	uint32 max_seq_length = 6;      // 最大序列长度，默认 2048
	uint32 save_steps = 7;          // 保存间隔步数，默认 100
	uint32 logging_steps = 8;       // 日志间隔步数，默认 10
}

// 数据集配置
message DatasetConfig {
	string path = 1;               // 数据集路径（JSON 文件或目录）
	string format = 2;             // 格式：jsonl, json, parquet
	double setting_oversample = 3; // 设定数据过采样倍数，默认 5.0
}

// 训练请求
message TrainingRequest {
	string job_id = 1;             // 任务 ID（客户端生成或服务端生成）
	string base_model = 2;         // 基础模型路径
	string output_dir = 3;         // 输出目录
	string character_name = 4;     // 角色名称（用于数据筛选）
	
	DatasetConfig dataset = 5;     // 数据集配置
	LoraConfig lora = 6;           // LoRA 配置
	QuantizationConfig quantization = 7;  // 量化配置
	TrainingHyperparams hyperparams = 8;  // 训练超参数
	
	StreamMetadata metadata = 9;   // 请求元数据
}

// 训练进度（流式返回）
message TrainingProgress {
	string job_id = 1;
	TrainingStatus status = 2;
	
	// 进度信息
	uint32 current_epoch = 3;
	uint32 total_epochs = 4;
	uint32 current_step = 5;
	uint32 total_steps = 6;
	double progress_percent = 7;
	
	// 训练指标
	double loss = 8;
	double learning_rate = 9;
	double gpu_memory_used = 10;   // GPU 显存使用（GB）
	double gpu_utilization = 11;  // GPU 利用率（%）
	
	// 时间信息
	double elapsed_seconds = 12;
	double eta_seconds = 13;       // 预计剩余时间
	
	// 消息
	string message = 14;
	string error = 15;
}

// 训练状态查询请求
message TrainingStatusRequest {
	string job_id = 1;
}

// 训练状态响应
message TrainingStatusResponse {
	string job_id = 1;
	TrainingStatus status = 2;
	TrainingProgress latest_progress = 3;
	string output_path = 4;        // 完成后的输出路径
}

// 取消训练请求
message CancelTrainingRequest {
	string job_id = 1;
}

// 取消训练响应
message CancelTrainingResponse {
	string job_id = 1;
	bool success = 2;
	string message = 3;
}

// 列出训练任务请求
message ListTrainingJobsRequest {
	repeated TrainingStatus status_filter = 1;  // 状态过滤
	uint32 limit = 2;              // 返回数量限制
}

// 训练任务摘要
message TrainingJobSummary {
	string job_id = 1;
	TrainingStatus status = 2;
	string base_model = 3;
	string character_name = 4;
	double progress_percent = 5;
	string created_at = 6;
	string updated_at = 7;
}

// 列出训练任务响应
message ListTrainingJobsResponse {
	repeated TrainingJobSummary jobs = 1;
}

// ========== 服务定义 ==========

// 推理服务
service AiInference {
	rpc StreamPredict(stream StreamPredictRequest) returns (stream StreamPredictResponse);
	rpc Control(ControlMessage) returns (ControlMessage);
	rpc Health(WorkerHealthRequest) returns (WorkerHealthResponse);
}

// 训练服务
service AiTraining {
	// 启动训练任务，流式返回进度
	rpc StartTraining(TrainingRequest) returns (stream TrainingProgress);
	
	// 查询训练状态
	rpc GetTrainingStatus(TrainingStatusRequest) returns (TrainingStatusResponse);
	
	// 取消训练任务
	rpc CancelTraining(CancelTrainingRequest) returns (CancelTrainingResponse);
	
	// 列出训练任务
	rpc ListTrainingJobs(ListTrainingJobsRequest) returns (ListTrainingJobsResponse);

	// 自定义脚本执行
	rpc ExecuteCustomScript(CustomScriptExecutionRequest) returns (stream CustomScriptExecutionProgress);

	// 取消自定义脚本
	rpc CancelCustomScript(CancelCustomScriptRequest) returns (CancelCustomScriptResponse);
}

// ====== 自定义脚本相关消息 ======
message CustomScriptExecutionRequest {
	string job_id = 1;
	string script_path = 2;
	repeated string args = 3;
	map<string, string> env = 4;
	string working_dir = 5;
	uint32 timeout_seconds = 6;
	StreamMetadata metadata = 7;
}

message CustomScriptExecutionProgress {
	string job_id = 1;
	string status = 2;
	repeated string stdout_lines = 3;
	repeated string stderr_lines = 4;
	optional int32 return_code = 5; // optional return code, use proto3 optional for hasReturnCode
	double elapsed_seconds = 6;
	map<string, string> metrics = 7; // metrics as strings (try to keep it simple)
	string error = 8;
}

message CancelCustomScriptRequest {
	string job_id = 1;
}

message CancelCustomScriptResponse {
	string job_id = 1;
	bool success = 2;
	string message = 3;
}
